### ISSUE TITLE
Swap array ordering on components

### ISSUE THREAD

**Issue by mcgibbon** [CliMT/climt#43]
One thing we've worried about a lot in working on CliMT is array ordering, especially C versus F ordering. After looking at the code, what we're currently doing is a little strange, and could be okay, but we should think about it and decide if we want to change it. Our two main considerations are that we want to minimize array copying, and we want things to work.

TL;DR: We should probably swap the dimension ordering on components.

How things currently work:

- We decided on a consistent dimension ordering among components, so that swapping dimensions between components isn't necessary. Currently that's `[horizontal, vertical]` or `[x, y, z]`, in C-ordering.
- The column-wise Fortran codes are generally written to take in arrays in `[horizontal, vertical]` (or `[column, vertical]`) order in F-ordering. This corresponds to `[vertical, horizontal]` in C-ordering. *If* we had any non-one number of columns passed into the Fortran codes, this would cause an error. However, the way we've written the wrappers, they pass in one column at a time to the Fortran code, so the ordering of the Fortran dimensions is actually irrelevant.

Some changes we might want to make:
- The code might run faster if we passed in multiple columns to the Fortran code at once, but I'm not sure. This would require testing. It would be the difference between applying a function F to each column of the model sequentially, and taking all columns and applying each step of the function F to all those columns at once. If we wanted to do this, we would need to switch CliMT's array ordering to be `[vertical, horizontal]`.
- I suppose we should probably just do whatever the dynamical core does? Which I think involves switching CliMT's array ordering to be `[vertical, horizontal]`. If we don't, we definitely need to copy when calling the dynamical core. If we switch the ordering for this reason, we *must* pass multiple columns at once to the Fortran codes, or we'd again have array copying.
- After switching, CliMT's Python and Fortran codes would use the same memory ordering arrays, [vertical, horizontal] in the C-ordered Python code, and [horizontal, vertical] in the F-ordered Fortran code.

**Comment by mcgibbon** [comment]
An additional consideration, however: If we switch the array ordering, we can no longer use prange to iterate over columns, because columns will no longer be contiguous in memory. At its core, this is because the dynamical core does not want columns to be contiguous in memory. I think the real solution to this is to look for a way to do parallelism other than prange (for example, by breaking the model into subdomains that are each running an MPI process that corresponds to the run script).

**Comment by JoyMonteiro** [comment]
Yes, this was something I have been wondering for some time now as well -- dycores split along the horizontal while physics requires vertical decomposition.

I'm with your suggestion to switch array ordering. However, I think this does not require us to
fully abandon prange, since if we have an array `temp[lon, lat, lev]` which takes in multiple levels, we could still prange on `lon`. However, that optimisation can come in later, and for now we could just pass in multiple columns.

**Comment by mcgibbon** [comment]
The issue is that we will have arrays `temp[lev, lat, lon]` or similar, with `lev` first as the slowest changing in-memory dimension, and you can't `prange` on `lev` (the call needs all levels). I'll alter things to use the new ordering and abandon `prange` where relevant.
